frontend/src/plugin/lib.ts

```diff
@@ -0,0 +1,141 @@
+ *
+ * @see Plugin
+ * @see registerPlugin
+ *
+ *
+ * ## Initialize Plugin(s)
+ *
+ * Each Plugin should have an initialize(registry) method.
+ * When they are initialized, they are given an Registry API object.
+ *
+ * @see Registry
+ *
+ *
+ * ## Functions available to plugins.
+ *
+ * ### APIs added to the browser window object.
+ *
+ * Some attributes are added to the browser "window" global for use by plugins.
+ *
+ * - window.registerPlugin, so plugins can register themselves.
+ * - window.plugins, a collection of plugins keyed by name.
+ * - window.pluginLib, some libraries exposed to use by plugins.
+ *
+ * Third party Libraries in window.pluginLib that can be used by plugins.
+ *
+ * - ReactRedux, {@link https://www.npmjs.com/package/@material-ui/core @material-ui/core}
+ * - MuiStyles, {@link https://www.npmjs.com/package/@material-ui/styles @material-ui/styles}
+ * - React, {@link https://www.npmjs.com/package/react react}
+ * - ReactRedux, {@link https://www.npmjs.com/package/react-redux react-redux}
+ * - Iconify, {@link https://www.npmjs.com/package/@iconify/react @iconify/react}
+ * - Lodash, {@link https://www.npmjs.com/package/lodash lodash}
+ *
+ * Headlamp libraries in window.pluginLib
+ *
+ * - CommonComponents, components/common, see `npm run storybook` in the headlamp/frontend repo.
+ * - K8s, frontend/src/lib/k8s
+ * - Utils frontend/src/lib/util
+ */
+import Registry from './registry';
+
+// @todo: types for CommonComponents, K8s, and Utils should be put into plugins-pkg.
+//        Because available in window.pluginLib.
+
+/**
+ * Plugins should call Headlamp.registerPlugin(pluginId: string, pluginObj: Plugin) to register themselves.
+ *
+ * They will have their initialize(register) method called at plugin initialization time.
+ */
+export abstract class Plugin {
+  /**
+   * initialize is called for each plugin with a Registry which gives the plugin methods for doing things.
+   *
+   * @see Registry
+   */
+  abstract initialize(register: Registry): boolean;
+}
+
+declare global {
+  interface Window {
+    pluginLib: {
+      [libName: string]: any;
+    };
+    plugins: {
+      [pluginId: string]: Plugin;
+    };
+    registerPlugin: (pluginId: string, pluginObj: Plugin) => void;
+  }
+}
+
+/**
+ * This class is a more convenient way for plugins to call registerPlugin in order to register
+ * themselves.
+ */
+export abstract class Headlamp {
+  /**
+   * Got a new plugin to add? Well, registerPlugin is your friend.
+   *
+   * @param pluginId - a unique id string for your plugin.
+   * @param pluginObj - the plugin being added.
+   *
+   * @example
+   *
+   * ```javascript
+   * const myPlugin = {
+   *   initialize: (register) => {
+   *     // do some stuff with register
+   *     // use some libraries in window.pluginLib
+   *     return true;
+   *   }
+   * }
+   *
+   * Headlamp.registerPlugin("aPluginIdString", myPlugin)
+   * ```
+   */
+  static registerPlugin(pluginId: string, pluginObj: Plugin) {
```

Comment:
Regarding [323](https://github.com/kinvolk/headlamp/issues/323#issuecomment-1020125006), I'm wondering if since we will change the API here... is the time to change the registerPlugin interface to address that problem? (Perhaps to use a react component rather than a plugin object with an initialize method).